submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,132 @@
+this enhancement.
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectivityPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec ClusterConnectivityPoliciesSpec `json:"spec"`
+}
+
+type ClusterConnectivityPoliciesSpec struct {
+    Policies []ClusterConnectivityPolicy `json:"policies,omitempty"`
+}
+
+type ClusterConnectivityPolicy struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    CableDriver *CableDriverSpec `json:"cableDriver,omitempty"`
+}
+
+type CableDriverSpec struct {
+    // Name of the cable driver implementation.
+    Name string `json:"name"`
+
+    // Options specifies key/value configuration parameters for the cable driver.
+    Options map[string]string `json:"options,omitempty"`
+}
+```
+
+The `ClusterConnectivityPolicies` CRD defines a list of policies that are applied in order of precedence. To resolve the
+connectivity policy for a remote cluster's `Endpoint`, Submariner would iterate the list of policies to search for a policy
+whose left and right label selectors match the corresponding `Cluster` resources of the local and remote `Endpoints`. The
+first matching policy is used. If no matching policy is found, the clusters are not connected.
+
+For full mesh connectivity, a policy with empty left and right label selectors would match all clusters. This would be the
+default.
+
+For each connectivity policy, the cable driver and associated configuration options can optionally be specified. If none
+is specified, the default cable driver is used.
+
+The `ClusterConnectivityPolicies` resource would be a singleton with a well-known name and would be maintained on the
+broker cluster and synced to each participating cluster.
+
+### Examples
+
+#### Full mesh (default)
+
+```yaml
+appVersion: v1
+kind: ClusterConnectivityPolicies
+metadata:
+  name: cluster-connectivity-policies
+spec:
+  policies:
+    - cableDriver:
+        name: libreswan
+```
+
+#### Use a specific cable driver to connect certain clusters
+
+The first policy specifies to use a specific cable driver to connect clusters labeled as on-premise followed by the fallback
+full mesh policy to connect everything else.
+
+```yaml
+appVersion: v1
+kind: ClusterConnectivityPolicies
+metadata:
+  name: cluster-connectivity-policies
+spec:
+  policies:
+    - leftClusterSelector:
+          location: on-premise
```

Comment:
Are the labels validated against `clustercrd`?

Reply:
yes - the submariner `Cluster` resource.